File: poc/game/src/systems/GameState.ts
// ============================================================
// Jurassic Trainers -- Game State (Singleton)
// ============================================================

import { Dino } from '../entities/Dino';
import { PartySystem } from './PartySystem';
import { InventorySystem } from './InventorySystem';
import { saveSystem, IGameState, IDinodexEntry } from './SaveSystem';
import { Direction } from '../utils/constants';

// ===================== Interfaces =====================

export interface IPlayerState {
  name: string;
  x: number;
  y: number;
  mapId: string;
  facing: Direction;
  money: number;
  badges: number[];
  playtime: number;   // seconds
}

// ===================== Game State Singleton =====================

export class GameState {
  private static instance: GameState | null = null;

  // --- Player ---
  player: IPlayerState;

  // --- Systems ---
  private partySystem: PartySystem;
  private inventorySystem: InventorySystem;

  // --- Collections ---
  dinodex: Map<number, IDinodexEntry>;
  flags: Map<string, boolean>;
  defeatedTrainers: Set<string>;

  // --- Runtime ---
  currentMap: string;
  private sessionStartTime: number;
  private accumulatedPlaytime: number;

  // --- Last heal point (respawn on blackout) ---
  lastHealMap: string;
  lastHealX: number;
  lastHealY: number;

  private constructor() {
    this.player = {
      name: '',
      x: 0,
      y: 0,
      mapId: 'BOURG_NID',
      facing: Direction.Down,
      money: 3000,
      badges: [],
      playtime: 0,
    };

    this.partySystem = new PartySystem();
    this.inventorySystem = new InventorySystem();
    this.dinodex = new Map();
    this.flags = new Map();
    this.defeatedTrainers = new Set();
    this.currentMap = 'BOURG_NID';
    this.lastHealMap = 'BOURG_NID';
    this.lastHealX = 7;
    this.lastHealY = 5;
    this.sessionStartTime = Date.now();
    this.accumulatedPlaytime = 0;
  }

  // --------------- Singleton ---------------

  static getInstance(): GameState {
    if (!GameState.instance) {
      GameState.instance = new GameState();
    }
    return GameState.instance;
  }

  /** Reset the singleton (for new game). */
  static reset(): void {
    GameState.instance = null;
  }

  // --------------- Initialization ---------------

  /**
   * Initialize a new game with player name and starter dino.
   */
  init(playerName: string, starterId: number): void {
    this.player.name = playerName;
    this.player.x = 7;   // Starting position (center of map)
    this.player.y = 5;
    this.player.mapId = 'BOURG_NID';
    this.player.facing = Direction.Down;
    this.player.money = 3000;
    this.player.badges = [];
    this.player.playtime = 0;

    this.partySystem = new PartySystem();
    this.inventorySystem = new InventorySystem();
    this.dinodex = new Map();
    this.flags = new Map();
    this.defeatedTrainers = new Set();
    this.currentMap = 'BOURG_NID';
    this.sessionStartTime = Date.now();
    this.accumulatedPlaytime = 0;

    // Create starter dino
    const starter = Dino.createStarter(starterId, 5);
    this.partySystem.addToParty(starter);
    this.catchDino(starterId);

    // Give starting items (IDs match items.json)
    this.inventorySystem.addItem(1, 3);   // 3 Potions (id=1)
    this.inventorySystem.addItem(16, 5);  // 5 Jurassic Balls (id=16)

    // Default heal point: starting village
    this.lastHealMap = 'BOURG_NID';
    this.lastHealX = 7;
    this.lastHealY = 5;
  }

  // --------------- System Accessors ---------------

  getPartySystem(): PartySystem {
    return this.partySystem;
  }

  getInventorySystem(): InventorySystem {
    return this.inventorySystem;
  }

  /** Shortcut: get the party array. */
  get party(): readonly Dino[] {
    return this.partySystem.getParty();
  }

  /** Shortcut: get the storage array. */
  get storage(): readonly Dino[] {
    return this.partySystem.getStorage();
  }

  // --------------- Dinodex ---------------

  seeDino(speciesId: number): void {
    const entry = this.dinodex.get(speciesId);
    if (entry) {
      entry.seen = true;
    } else {
      this.dinodex.set(speciesId, { seen: true, caught: false });
    }
  }

  catchDino(speciesId: number): void {
    const entry = this.dinodex.get(speciesId);
    if (entry) {
      entry.seen = true;
      entry.caught = true;
    } else {
      this.dinodex.set(speciesId, { seen: true, caught: true });
    }
  }

  getDinodexEntry(speciesId: number): IDinodexEntry | undefined {
    return this.dinodex.get(speciesId);
  }

  getDinodexSeenCount(): number {
    let count = 0;
    for (const entry of this.dinodex.values()) {
      if (entry.seen) count++;
    }
    return count;
  }

  getDinodexCaughtCount(): number {
    let count = 0;
    for (const entry of this.dinodex.values()) {
      if (entry.caught) count++;
    }
    return count;
  }

  // --------------- Badges ---------------

  addBadge(index: number): void {
    if (!this.player.badges.includes(index)) {
      this.player.badges.push(index);
      this.player.badges.sort((a, b) => a - b);
    }
  }

  hasBadge(index: number): boolean {
    return this.player.badges.includes(index);
  }

  getBadgeCount(): number {
    return this.player.badges.length;
  }

  // --------------- Heal Point (respawn) ---------------

  /** Save the current position as the respawn point (call when healing at a center) */
  setHealPoint(mapId: string, x: number, y: number): void {
    this.lastHealMap = mapId;
    this.lastHealX = x;
    this.lastHealY = y;
  }

  /** Get the last heal point for respawn on blackout */
  getHealPoint(): { mapId: string; x: number; y: number } {
    return { mapId: this.lastHealMap, x: this.lastHealX, y: this.lastHealY };
  }

  // --------------- Flags ---------------

  setFlag(key: string, value: boolean = true): void {
    this.flags.set(key, value);
  }

  hasFlag(key: string): boolean {
    return this.flags.get(key) === true;
  }

  getFlag(key: string): boolean {
    return this.flags.get(key) ?? false;
  }

  // --------------- Trainers ---------------

  defeatTrainer(id: string): void {
    this.defeatedTrainers.add(id);
  }

  isTrainerDefeated(id: string): boolean {
    return this.defeatedTrainers.has(id);
  }

  // --------------- Money ---------------

  addMoney(amount: number): void {
    this.player.money += Math.max(0, Math.floor(amount));
    // Cap at 999,999
    this.player.money = Math.min(this.player.money, 999999);
  }

  removeMoney(amount: number): boolean {
    const cost = Math.max(0, Math.floor(amount));
    if (this.player.money < cost) return false;
    this.player.money -= cost;
    return true;
  }

  getMoney(): number {
    return this.player.money;
  }

  // --------------- Playtime ---------------

  /** Get total playtime in seconds (accumulated + current session). */
  getPlaytime(): number {
    const sessionSeconds = Math.floor((Date.now() - this.sessionStartTime) / 1000);
    return this.accumulatedPlaytime + sessionSeconds;
  }

  // --------------- Position ---------------

  setPlayerPosition(x: number, y: number, mapId?: string): void {
    this.player.x = x;
    this.player.y = y;
    if (mapId) {
      this.player.mapId = mapId;
      this.currentMap = mapId;
    }
  }

  setPlayerFacing(direction: Direction): void {
    this.player.facing = direction;
  }

  // --------------- Save / Load ---------------

  /**
   * Save the current game state to a slot.
   */
  saveGame(slot: number): boolean {
    const state = this.serializeState();
    return saveSystem.save(slot, state);
  }

  /**
   * Load a game state from a slot and apply it.
   */
  loadGame(slot: number): boolean {
    const state = saveSystem.load(slot);
    if (!state) return false;
    this.deserializeState(state);
    return true;
  }

  private serializeState(): IGameState {
    // Convert dinodex Map to Record
    const dinodexRecord: Record<number, IDinodexEntry> = {};
    for (const [id, entry] of this.dinodex) {
      dinodexRecord[id] = { ...entry };
    }

    // Convert flags Map to Record
    const flagsRecord: Record<string, boolean> = {};
    for (const [key, value] of this.flags) {
      flagsRecord[key] = value;
    }

    return {
      version: 1,
      player: {
        name: this.player.name,
        x: this.player.x,
        y: this.player.y,
        mapId: this.player.mapId,
        facing: this.player.facing,
      },
      party: this.partySystem.serializeParty(),
      storage: this.partySystem.serializeStorage(),
      inventory: this.inventorySystem.serialize(),
      dinodex: dinodexRecord,
      badges: [...this.player.badges],
      flags: flagsRecord,
      defeatedTrainers: Array.from(this.defeatedTrainers),
      playtime: this.getPlaytime(),
      money: this.player.money,
      savedAt: new Date().toISOString(),
    };
  }

  private deserializeState(state: IGameState): void {
    // Player
    this.player.name = state.player.name;
    this.player.x = state.player.x;
    this.player.y = state.player.y;
    this.player.mapId = state.player.mapId;
    this.player.facing = state.player.facing as Direction;
    this.player.money = state.money;
    this.player.badges = [...state.badges];
    this.currentMap = state.player.mapId;

    // Playtime
    this.accumulatedPlaytime = state.playtime;
    this.sessionStartTime = Date.now();

    // Party & Storage
    this.partySystem = new PartySystem();
    this.partySystem.loadParty(state.party);
    this.partySystem.loadStorage(state.storage);

    // Inventory
    this.inventorySystem = InventorySystem.deserialize(state.inventory);

    // Dinodex
    this.dinodex = new Map();
    for (const [idStr, entry] of Object.entries(state.dinodex)) {
      this.dinodex.set(Number(idStr), { ...entry });
    }

    // Flags
    this.flags = new Map();
    for (const [key, value] of Object.entries(state.flags)) {
      this.flags.set(key, value);
    }

    // Defeated trainers
    this.defeatedTrainers = new Set(state.defeatedTrainers);
  }
}

File: DinoMonsters/Assets/Scripts/Core/GameState.cs
// ============================================================
// Dino Monsters -- Game State (Plain C# Singleton)
// Ported from TypeScript POC to C# (Unity)
// ============================================================

using System;
using System.Collections.Generic;
using UnityEngine;

// --------------- Serializable Sub-Structures ---------------

[System.Serializable]
public class HealPoint
{
    public string mapId;
    public int x;
    public int y;

    public HealPoint()
    {
        mapId = "BOURG_NID";
        x = 7;
        y = 5;
    }

    public HealPoint(string mapId, int x, int y)
    {
        this.mapId = mapId;
        this.x = x;
        this.y = y;
    }
}

[System.Serializable]
public class DinodexEntry
{
    public int speciesId;
    public bool seen;
    public bool caught;
}

[System.Serializable]
public class FlagEntry
{
    public string key;
    public bool value;
}

// InventoryEntry is defined in InventorySystem.cs

// --------------- Serializable Save Data ---------------

[System.Serializable]
public class GameSaveData
{
    public int version = 1;
    public string playerName;
    public int money;
    public bool[] badges = new bool[8];
    public List<Dino> party = new List<Dino>();
    public List<Dino> pcStorage = new List<Dino>();
    public List<InventoryEntry> inventory = new List<InventoryEntry>();
    public List<DinodexEntry> dinodex = new List<DinodexEntry>();
    public List<FlagEntry> flags = new List<FlagEntry>();
    public List<string> defeatedTrainers = new List<string>();
    public HealPoint healPoint = new HealPoint();
    public string currentMapId;
    public int playerX;
    public int playerY;
    public float playtime;
    public string savedAt;
}

// --------------- Game State Singleton ---------------

public class GameState
{
    private static GameState instance;

    // --- Player ---
    public string PlayerName { get; set; }
    public int Money { get; private set; }
    public bool[] Badges { get; private set; }

    // --- Party & Storage ---
    public List<Dino> Party { get; private set; }
    public List<Dino> PCStorage { get; private set; }
    public const int MAX_PARTY_SIZE = 6;

    // --- Inventory ---
    public InventorySystem Inventory { get; private set; }

    // --- Story Progression ---
    private Dictionary<string, bool> flags;
    private HashSet<string> defeatedTrainers;

    // --- Dinodex ---
    private Dictionary<int, DinodexEntry> dinodex;

    // --- Location ---
    public string CurrentMapId { get; set; }
    public int PlayerX { get; set; }
    public int PlayerY { get; set; }
    public HealPoint LastHealPoint { get; private set; }

    // --- Playtime ---
    private float sessionStartTime;
    private float accumulatedPlaytime;

    // --------------- Singleton ---------------

    public static GameState Instance
    {
        get
        {
            if (instance == null)
                instance = new GameState();
            return instance;
        }
    }

    public static void Reset()
    {
        instance = null;
    }

    // --------------- Constructor ---------------

    private GameState()
    {
        PlayerName = "";
        Money = 3000;
        Badges = new bool[8];
        Party = new List<Dino>();
        PCStorage = new List<Dino>();
        Inventory = new InventorySystem();
        flags = new Dictionary<string, bool>();
        defeatedTrainers = new HashSet<string>();
        dinodex = new Dictionary<int, DinodexEntry>();
        CurrentMapId = "BOURG_NID";
        PlayerX = 7;
        PlayerY = 5;
        LastHealPoint = new HealPoint();
        sessionStartTime = Time.realtimeSinceStartup;
        accumulatedPlaytime = 0f;
    }

    // --------------- Initialization ---------------

    public void Init(string playerName, int starterId)
    {
        PlayerName = playerName;
        Money = 3000;
        Badges = new bool[8];
        Party = new List<Dino>();
        PCStorage = new List<Dino>();
        Inventory = new InventorySystem();
        flags = new Dictionary<string, bool>();
        defeatedTrainers = new HashSet<string>();
        dinodex = new Dictionary<int, DinodexEntry>();
        CurrentMapId = "BOURG_NID";
        PlayerX = 7;
        PlayerY = 5;
        LastHealPoint = new HealPoint();
        sessionStartTime = Time.realtimeSinceStartup;
        accumulatedPlaytime = 0f;

        // Create starter
        var starter = Dino.CreateStarter(starterId, 5);
        AddToParty(starter);
        RegisterCaught(starterId);

        // Starting items
        Inventory.AddItem(1, 3);   // 3 Potions
        Inventory.AddItem(16, 5);  // 5 Jurassic Balls
    }

    // --------------- Party Management ---------------

    public bool AddToParty(Dino dino)
    {
        if (Party.Count >= MAX_PARTY_SIZE) return false;
        Party.Add(dino);
        return true;
    }

    public void AddToPC(Dino dino)
    {
        PCStorage.Add(dino);
    }

    public Dino GetLeadDino()
    {
        return Party.Count > 0 ? Party[0] : null;
    }

    public bool HasAliveDino()
    {
        foreach (var dino in Party)
        {
            if (!dino.IsFainted()) return true;
        }
        return false;
    }

    // --------------- Money ---------------

    public void AddMoney(int amount)
    {
        Money = Mathf.Min(Money + Mathf.Max(0, amount), 999999);
    }

    public bool RemoveMoney(int amount)
    {
        int cost = Mathf.Max(0, amount);
        if (Money < cost) return false;
        Money -= cost;
        return true;
    }

    // --------------- Badges ---------------

    public void AddBadge(int index)
    {
        if (index >= 0 && index < 8)
            Badges[index] = true;
    }

    public bool HasBadge(int index)
    {
        if (index < 0 || index >= 8) return false;
        return Badges[index];
    }

    public int GetBadgeCount()
    {
        int count = 0;
        foreach (bool b in Badges)
        {
            if (b) count++;
        }
        return count;
    }

    // --------------- Flags ---------------

    public void SetFlag(string key, bool value = true)
    {
        flags[key] = value;
    }

    public bool HasFlag(string key)
    {
        return flags.ContainsKey(key) && flags[key];
    }

    public bool GetFlag(string key)
    {
        return flags.ContainsKey(key) && flags[key];
    }

    // --------------- Trainers ---------------

    public void DefeatTrainer(string id)
    {
        defeatedTrainers.Add(id);
    }

    public bool IsTrainerDefeated(string id)
    {
        return defeatedTrainers.Contains(id);
    }

    // --------------- Heal Point ---------------

    public void SetHealPoint(string mapId, int x, int y)
    {
        LastHealPoint = new HealPoint(mapId, x, y);
    }

    // --------------- Dinodex ---------------

    public void RegisterSeen(int speciesId)
    {
        if (!dinodex.ContainsKey(speciesId))
        {
            dinodex[speciesId] = new DinodexEntry { speciesId = speciesId, seen = true, caught = false };
        }
        else
        {
            dinodex[speciesId].seen = true;
        }
    }

    public void RegisterCaught(int speciesId)
    {
        if (!dinodex.ContainsKey(speciesId))
        {
            dinodex[speciesId] = new DinodexEntry { speciesId = speciesId, seen = true, caught = true };
        }
        else
        {
            dinodex[speciesId].seen = true;
            dinodex[speciesId].caught = true;
        }
    }

    public DinodexEntry GetDinodexEntry(int speciesId)
    {
        return dinodex.ContainsKey(speciesId) ? dinodex[speciesId] : null;
    }

    public int GetDinodexSeenCount()
    {
        int count = 0;
        foreach (var entry in dinodex.Values)
        {
            if (entry.seen) count++;
        }
        return count;
    }

    public int GetDinodexCaughtCount()
    {
        int count = 0;
        foreach (var entry in dinodex.Values)
        {
            if (entry.caught) count++;
        }
        return count;
    }

    // --------------- Playtime ---------------

    public float GetPlaytime()
    {
        float sessionSeconds = Time.realtimeSinceStartup - sessionStartTime;
        return accumulatedPlaytime + sessionSeconds;
    }

    // --------------- Save / Load ---------------

    public GameSaveData Serialize()
    {
        var data = new GameSaveData();
        data.version = 1;
        data.playerName = PlayerName;
        data.money = Money;
        data.badges = (bool[])Badges.Clone();
        data.party = new List<Dino>(Party);
        data.pcStorage = new List<Dino>(PCStorage);
        data.currentMapId = CurrentMapId;
        data.playerX = PlayerX;
        data.playerY = PlayerY;
        data.healPoint = new HealPoint(LastHealPoint.mapId, LastHealPoint.x, LastHealPoint.y);
        data.playtime = GetPlaytime();
        data.savedAt = DateTime.Now.ToString("o");

        // Inventory
        data.inventory = Inventory.SerializeToList();

        // Dinodex
        data.dinodex = new List<DinodexEntry>();
        foreach (var entry in dinodex.Values)
        {
            data.dinodex.Add(entry);
        }

        // Flags
        data.flags = new List<FlagEntry>();
        foreach (var kvp in flags)
        {
            data.flags.Add(new FlagEntry { key = kvp.Key, value = kvp.Value });
        }

        // Defeated trainers
        data.defeatedTrainers = new List<string>(defeatedTrainers);

        return data;
    }

    public void Deserialize(GameSaveData data)
    {
        PlayerName = data.playerName;
        Money = data.money;
        Badges = (bool[])data.badges.Clone();
        CurrentMapId = data.currentMapId;
        PlayerX = data.playerX;
        PlayerY = data.playerY;
        LastHealPoint = new HealPoint(data.healPoint.mapId, data.healPoint.x, data.healPoint.y);
        accumulatedPlaytime = data.playtime;
        sessionStartTime = Time.realtimeSinceStartup;

        // Party — recalculate stats from save
        Party = new List<Dino>();
        foreach (var dino in data.party)
        {
            Party.Add(Dino.CreateFromSerialized(dino));
        }

        // PC Storage
        PCStorage = new List<Dino>();
        foreach (var dino in data.pcStorage)
        {
            PCStorage.Add(Dino.CreateFromSerialized(dino));
        }

        // Inventory
        Inventory = new InventorySystem();
        Inventory.DeserializeFromList(data.inventory);

        // Dinodex
        dinodex = new Dictionary<int, DinodexEntry>();
        foreach (var entry in data.dinodex)
        {
            dinodex[entry.speciesId] = entry;
        }

        // Flags
        flags = new Dictionary<string, bool>();
        foreach (var entry in data.flags)
        {
            flags[entry.key] = entry.value;
        }

        // Defeated trainers
        defeatedTrainers = new HashSet<string>(data.defeatedTrainers);
    }
}
